File: metadata/mill-metadata-core/src/test/java/io/qpointz/mill/metadata/domain/FacetTypeDescriptorTest.java
package io.qpointz.mill.metadata.domain;

import org.junit.jupiter.api.Test;

import java.util.Set;

import static org.junit.jupiter.api.Assertions.*;

class FacetTypeDescriptorTest {

    @Test
    void shouldBeApplicableToAny_whenApplicableToIsNull() {
        var descriptor = FacetTypeDescriptor.builder()
                .typeKey("test")
                .applicableTo(null)
                .build();

        assertTrue(descriptor.isApplicableTo(MetadataTargetType.TABLE));
        assertTrue(descriptor.isApplicableTo(MetadataTargetType.CONCEPT));
    }

    @Test
    void shouldBeApplicableToAny_whenApplicableToIsEmpty() {
        var descriptor = FacetTypeDescriptor.builder()
                .typeKey("test")
                .applicableTo(Set.of())
                .build();

        assertTrue(descriptor.isApplicableTo(MetadataTargetType.TABLE));
    }

    @Test
    void shouldBeApplicableToAny_whenContainsANY() {
        var descriptor = FacetTypeDescriptor.builder()
                .typeKey("test")
                .applicableTo(Set.of(MetadataTargetType.ANY))
                .build();

        assertTrue(descriptor.isApplicableTo(MetadataTargetType.TABLE));
        assertTrue(descriptor.isApplicableTo(MetadataTargetType.CONCEPT));
    }

    @Test
    void shouldRestrict_whenApplicableToIsSpecific() {
        var descriptor = FacetTypeDescriptor.builder()
                .typeKey("structural")
                .applicableTo(Set.of(MetadataTargetType.TABLE, MetadataTargetType.ATTRIBUTE))
                .build();

        assertTrue(descriptor.isApplicableTo(MetadataTargetType.TABLE));
        assertTrue(descriptor.isApplicableTo(MetadataTargetType.ATTRIBUTE));
        assertFalse(descriptor.isApplicableTo(MetadataTargetType.CONCEPT));
        assertFalse(descriptor.isApplicableTo(MetadataTargetType.SCHEMA));
    }

    @Test
    void shouldDetectContentSchema() {
        var withSchema = FacetTypeDescriptor.builder()
                .typeKey("test")
                .contentSchema(java.util.Map.of("type", "object"))
                .build();
        assertTrue(withSchema.hasContentSchema());

        var withoutSchema = FacetTypeDescriptor.builder()
                .typeKey("test")
                .build();
        assertFalse(withoutSchema.hasContentSchema());

        var emptySchema = FacetTypeDescriptor.builder()
                .typeKey("test")
                .contentSchema(java.util.Map.of())
                .build();
        assertFalse(emptySchema.hasContentSchema());
    }

    @Test
    void shouldHaveCorrectDefaults() {
        var descriptor = FacetTypeDescriptor.builder()
                .typeKey("test")
                .build();

        assertFalse(descriptor.isMandatory());
        assertTrue(descriptor.isEnabled());
    }
}

File: metadata/mill-metadata-core/src/test/kotlin/io/qpointz/mill/metadata/domain/FacetTypeDescriptorTest.kt
package io.qpointz.mill.metadata.domain

import org.junit.jupiter.api.Assertions.assertFalse
import org.junit.jupiter.api.Assertions.assertTrue
import org.junit.jupiter.api.Test

class FacetTypeDescriptorTest {

    @Test
    fun shouldBeApplicableToAny_whenApplicableToIsNull() {
        val descriptor = FacetTypeDescriptor(typeKey = "test", applicableTo = null)
        assertTrue(descriptor.isApplicableTo(MetadataUrns.ENTITY_TYPE_TABLE))
        assertTrue(descriptor.isApplicableTo(MetadataUrns.ENTITY_TYPE_CONCEPT))
    }

    @Test
    fun shouldBeApplicableToAny_whenApplicableToIsEmpty() {
        val descriptor = FacetTypeDescriptor(typeKey = "test", applicableTo = emptySet())
        assertTrue(descriptor.isApplicableTo(MetadataUrns.ENTITY_TYPE_TABLE))
    }

    @Test
    fun shouldRestrict_whenApplicableToIsSpecific() {
        val descriptor = FacetTypeDescriptor(
            typeKey = "structural",
            applicableTo = setOf(MetadataUrns.ENTITY_TYPE_TABLE, MetadataUrns.ENTITY_TYPE_ATTRIBUTE)
        )
        assertTrue(descriptor.isApplicableTo(MetadataUrns.ENTITY_TYPE_TABLE))
        assertTrue(descriptor.isApplicableTo(MetadataUrns.ENTITY_TYPE_ATTRIBUTE))
        assertFalse(descriptor.isApplicableTo(MetadataUrns.ENTITY_TYPE_CONCEPT))
        assertFalse(descriptor.isApplicableTo(MetadataUrns.ENTITY_TYPE_SCHEMA))
    }

    @Test
    fun shouldDetectContentSchema() {
        val withSchema = FacetTypeDescriptor(typeKey = "test", contentSchema = mapOf("type" to "object"))
        assertTrue(withSchema.hasContentSchema())
        val withoutSchema = FacetTypeDescriptor(typeKey = "test")
        assertFalse(withoutSchema.hasContentSchema())
        val emptySchema = FacetTypeDescriptor(typeKey = "test", contentSchema = emptyMap())
        assertFalse(emptySchema.hasContentSchema())
    }

    @Test
    fun shouldHaveCorrectDefaults() {
        val descriptor = FacetTypeDescriptor(typeKey = "test")
        assertFalse(descriptor.mandatory)
        assertTrue(descriptor.enabled)
    }
}
